src/Cake.SqlServer/Backup/RestoreSqlBackupImpl.cs

```diff
@@ -11,65 +12,107 @@ internal static class RestoreSqlBackupImpl
+                    differentialCommand.ExecuteNonQuery();
+                }
+            }
+        }
 
-                var logicalNames = GetLogicalNames(backupFile, connection);
+        private static SqlCommand GetSqlCommand(ICakeContext context, SqlConnection connection, RestoreSqlBackupSettings settings, IList<FilePath> backupFiles, int backupSetFile = 1, bool noRecovery = false, bool allowAlterDb = true)
+        {
+            var firstBackupFile = backupFiles.First();
+            var oldDbName = GetDatabaseName(firstBackupFile, connection);
+            var newDatabaseName = settings.NewDatabaseName ?? oldDbName;
+            context.Log.Information($"Using database name '{newDatabaseName}' to be a name for the restored database");
 
-                var sql = "";
+            var logicalNames = GetLogicalNames(firstBackupFile, connection);
 
-                if (settings.SwitchToSingleUserMode)
-                {
-                    sql += $@"
+            var sql = "";
+
+            if (settings.SwitchToSingleUserMode && allowAlterDb)
+            {
+                sql += $@"
 if db_id({Sql.EscapeNameQuotes(newDatabaseName)}) is not null 
 begin
     use master;
     alter database {Sql.EscapeName(newDatabaseName)} set single_user with rollback immediate;
 end
 ";
+            }
+
+            sql += $"Restore database {Sql.EscapeName(newDatabaseName)} from\r\n";
+            for (var i = 0; i < backupFiles.Count; i++)
+            {
+                sql += $" disk = @backupFile{i}";
+                if (i < backupFiles.Count - 1)
+                {
+                    sql += ","; // only need comma before penultimate list
                 }
 
-                sql += $"Restore database {Sql.EscapeName(newDatabaseName)} from disk = @backupFile with ";
+                sql += "\r\n";
+            }
 
-                for (var i = 0; i < logicalNames.Count; i++)
+            sql += $" with file = {backupSetFile},\r\n";
+            for (var i = 0; i < logicalNames.Count; i++)
+            {
+                sql += $" move @LName{i} to @LPath{i}";
+                if (i < logicalNames.Count - 1)
                 {
-                    sql += $" move @LName{i} to @LPath{i}";
-                    if (i < logicalNames.Count - 1)
-                    {
-                        sql += ", \r\n"; // only need comma before penultimate list
-                    }
+                    sql += ",\r\n"; // only need comma before penultimate list
                 }
-                if (settings.WithReplace)
+            }
+
+            if (settings.WithReplace)
+            {
+                sql += ",\r\n replace";
+            }
+
+            if (noRecovery)
+            {
+                sql += ",\r\n norecovery;";
+            }
+            else
+            {
+                sql += $",\r\n recovery;";
+                if (allowAlterDb)
```

Comment:
After some thought I believe it's ok to have several parameters for GetRestoreSqlBackupCommand because a single restore database command can (in general) have several options. However it should definitely have only that responsibility (in particular it shouldn't manage argument parsing or control logic).

All argument decisions and control logic is now done in RestoreSqlBackup.